lib/secretsManager.js

```diff
@@ -1,77 +1,154 @@
-let DUMMY_UUID_GEN = 0
-let DUMMY_STORE = []
-
-/**
- * Find all secrets, applying a selector if provided.
- * When provided, all where clause must be fulfilled (logical AND).
- * By default, returns all secrets
- * @async
- * @param {Object} where - hash of where clauses returned secrets must fulfill with
- * @returns {Array<Object>} list (may be empty) of matching secrets
- */
-export async function find(where = {}) {
-  let secrets = []
-
-  try {
-    const props = Object.keys(where)
-    secrets = DUMMY_STORE.filter(secret =>
-      props.every(prop => secret[prop] === where[prop])
+import uuid from 'uuid-random'
+import aesjs from 'aes-js'
+import * as SecureStore from 'expo-secure-store'
+import AsyncStorage from '@react-native-async-storage/async-storage'
+
+const encryptionPassphraseStorageKey = 'encryptionPassphrase'
+const secretStorageKey = 'secrets'
+
+export default class SecretsManager {
+  constructor() {
+    this.encryptionPassphrase = null
+  }
+
+  async init() {
+    let getEncryptionKey = AsyncStorage.getItem
+    let setEncryptionKey = AsyncStorage.setItem
+
+    if (SecureStore.isAvailableAsync()) {
+      getEncryptionKey = SecureStore.getItemAsync
+      setEncryptionKey = SecureStore.setItemAsync
+    }
+
+    const encryptionPassphrase = await getEncryptionKey(
+      encryptionPassphraseStorageKey
     )
-  } catch (err) {
-    console.error('Failed to read secrets from data store', err)
+
+    if (encryptionPassphrase) {
+      this.encryptionPassphrase = encryptionPassphrase
+      return
+    }
+
+    this.encryptionPassphrase = this.generateEncryptionPassphrase()
+
+    await setEncryptionKey(
+      encryptionPassphraseStorageKey,
+      this.encryptionPassphrase
+    )
+
+    await this.set([])
   }
 
-  return secrets
-}
+  generateEncryptionPassphrase() {
+    return JSON.stringify(
+      Array.from({ length: 32 }).map(() => Math.floor(Math.random() * 256))
+    )
```

Comment:
this is not a cryptographic secure way to generate a random key. I appreciate this is probably not a big deal in this specific context, but since we're following good cryptographic patterns, let's just follow them completely. you need to find something that's equivalent to the Node.js crypto library.

Reply:
I have to admit I'm puzzled by this task. Here I just kind of followed the docs at https://www.npmjs.com/package/aes-js#keys. I explored a bit with the possibility of using https://www.npmjs.com/package/pbkdf2, https://www.npmjs.com/package/scrypt-js or https://www.npmjs.com/package/bcryptjs but I couldn't quite figure how to put one of them together with aes-js. pbkdf2 for instance appears to rely on the native nodejs crypto lib and I couldn't get it to work on a first try. bcryptjs and scrypt-js both demand some kind of a input to be turned into a hash and then their output must be a char string of 16, 24 or 32 bytes to be accepted by the aes-js aes cipher functions which I couldn't figure out exactly as well on a first try. https://www.npmjs.com/package/aes-js#ctr---counter-recommended is the example I followed to implement this iteration.

Reply:
The approach you followed makes sense, the only bit that's missing, which as I said is probably a relatively minor detail in these specific circumstances even though it would be a pity if we missed this step, is that the key you use should be cryptocraphically secure. In Node.js you would use crypto.randomBytes to generate such a key, but using Math.random is not the same thing.

Now, because in AES the keys need to be of a specific length, you have 2 options:

- generate a truly random key of the required length
- use a password-based key derivation function which, given a password string of random length and not necessarily cryptographically secure, gives you a cryptographically secure key (script, bcrypt, pdkdf2)

Clearly, in option 2, you still have the issue of generating that password, but luckily we don't need that, because there's an extremely popular library which does the same as Node's randomBytes for react native: https://www.npmjs.com/package/react-native-get-random-values. This also seems to work fine in expo, so that should do

Reply:
I updated the code with a new version that uses the `react-native-get-random-values` lib. Thanks for clearing things out a bit for me here.